api/src/main/java/org/openmrs/module/oauth2/api/smart/impl/SmartAppLaunchServiceImpl.java

```diff
@@ -0,0 +1,67 @@
+package org.openmrs.module.oauth2.api.smart.impl;
+
+import java.math.BigInteger;
+import java.security.SecureRandom;
+
+import org.openmrs.api.impl.BaseOpenmrsService;
+import org.openmrs.module.oauth2.api.smart.SmartAppLaunchService;
+import org.openmrs.module.oauth2.api.smart.db.hibernate.LaunchValueDAO;
+import org.openmrs.module.oauth2.api.smart.model.LaunchValue;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.transaction.annotation.Transactional;
+
+public class SmartAppLaunchServiceImpl extends BaseOpenmrsService implements SmartAppLaunchService {
+
+	@Autowired
+	LaunchValueDAO launchValueDAO;
+
+	@Transactional
+	@Override
+	public LaunchValue saveOrUpdateLaunchValue(LaunchValue launchValue) {
+		return launchValueDAO.saveOrUpdate(launchValue);
+	}
+
+	@Transactional
+	@Override
+	public LaunchValue updateLaunchValue(LaunchValue launchValue) {
+		return launchValueDAO.update(launchValue);
+	}
+
+	@Transactional(readOnly = true)
+	@Override
+	public LaunchValue getLaunchValue(Integer smartId) {
+		return launchValueDAO.getById(smartId);
+	}
+
+	@Transactional
+	@Override
+	public LaunchValue clearLaunchValue(LaunchValue launchValue) {
+		launchValueDAO.delete(launchValue);
+		return launchValue;
+	}
+
+	@Transactional
+	@Override
+	public LaunchValue merge(LaunchValue launchValue) {
+		return launchValueDAO.merge(launchValue);
+	}
+
+	@Override
+	public void generateAndPersistLaunchValue(LaunchValue launchValue) {
+		generateLaunchValue(launchValue);
+		merge(launchValue);
+		updateLaunchValue(launchValue);
+	}
+
+	private void generateLaunchValue(LaunchValue launchValue) {
+		SecureRandom secureRandom = new SecureRandom();
+		launchValue.setLaunchValue(new BigInteger(130, secureRandom).toString(32));
+	}
+
+	@Override
+	public boolean verifyLaunchValue(LaunchValue launchValue, String receivedLaunchValue) {
+		if (launchValue.getLaunchValue().equals(receivedLaunchValue))
+			return true;
+		return false;
+	}
```

Comment:
When a new smartApp is invoked from OMRS, a launchValue will be generated. This launchValue will be passed back to Oauth2 module for verification. How'd you verify that the launch value is of the same smartApp if you don't verify it? There should be a method to get smartApp from launchValue right?

Reply:
Because, only looking at the launchValue we can see if its from a smartApp or not. i.e. do we need to process it like a normal oauth2_client or a smartApp

Reply:
I think the verifyLaunchValue() method will do the job?